airflow/dag/serialization.py

```diff
@@ -0,0 +1,316 @@
+            elif isinstance(x, dict):
+                return cls.encode(
+                    {str(k): cls._serialize(v, visited_dags)
+                     for k, v in x.items()},
+                    type_=DagTypes.DICT)
+            elif isinstance(x, list):
+                return [cls._serialize(v, visited_dags) for v in x]
+            elif isinstance(x, DAG):
+                return SerializedDAG.serialize_dag(x, visited_dags)
+            elif isinstance(x, BaseOperator):
+                return SerializedOperator.serialize_operator(x, visited_dags)
+            elif isinstance(x, cls._datetime_types):
+                return cls.encode(x.isoformat(), type_=DagTypes.DATETIME)
+            elif isinstance(x, datetime.timedelta):
+                return cls.encode(x.total_seconds(), type_=DagTypes.TIMEDELTA)
+            elif isinstance(x, pendulum.tz.Timezone):
+                return cls.encode(str(x.name), type_=DagTypes.TIMEZONE)
+            elif callable(x):
+                return str(get_python_source(x))
+            elif isinstance(x, set):
+                return cls.encode(
+                    [cls._serialize(v, visited_dags) for v in x], type_=DagTypes.SET)
+            elif isinstance(x, tuple):
+                return cls.encode(
+                    [cls._serialize(v, visited_dags) for v in x], type_=DagTypes.TUPLE)
+            else:
+                logging.debug('Cast type %s to str in serialization.', type(x))
+                return str(x)
+        except Exception:  # pylint: disable=broad-except
+            logging.warning('Failed to stringify.', exc_info=True)
+            return FAILED
+
+    @classmethod
+    def _deserialize(cls, encoded_x, visited_dags):  # pylint: disable=too-many-return-statements
+        """Helper function of depth first search for deserialization."""
+        try:
+            # JSON primitives (except for dict) are not encoded.
+            if cls._is_primitive(encoded_x):
+                return encoded_x
+            elif isinstance(encoded_x, list):
+                return [cls._deserialize(v, visited_dags) for v in encoded_x]
+
+            assert isinstance(encoded_x, dict)
+            var = encoded_x[Encoding.VAR]
+            type_ = encoded_x[Encoding.TYPE]
+
+            if type_ == DagTypes.DICT:
+                return {k: cls._deserialize(v, visited_dags) for k, v in var.items()}
+            elif type_ == DagTypes.DAG:
+                if isinstance(var, dict):
+                    return SerializedDAG.deserialize_dag(var, visited_dags)
+                elif isinstance(var, str) and var in visited_dags:
+                    # dag_id is stored in the serailized form for a visited DAGs.
+                    return visited_dags[var]
+                logging.warning('Invalid DAG %s in deserialization.', var)
+                return None
+            elif type_ == DagTypes.OP:
+                return SerializedOperator.deserialize_operator(var, visited_dags)
```

Comment:
I'm not sure about the naming here `SerializedOperator` might also sound like just another operator.

Reply:
Changed it to SerializedBaseOperator.